pkg/controller/certificatesigningrequests/vault/vault.go

```diff
@@ -0,0 +1,159 @@
+func (v *Vault) Sign(ctx context.Context, csr *certificatesv1.CertificateSigningRequest, issuerObj cmapi.GenericIssuer) error {
+	log := logf.FromContext(ctx, "sign")
+	log = logf.WithRelatedResource(log, issuerObj)
+
+	resourceNamespace := v.issuerOptions.ResourceNamespace(issuerObj)
+
+	client, err := v.clientBuilder(resourceNamespace, v.secretsLister, issuerObj)
+	if apierrors.IsNotFound(err) {
+		message := "Required secret resource not found"
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "SecretNotFound", message)
+		util.CertificateSigningRequestSetFailed(csr, "SecretNotFound", message)
+		_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	if err != nil {
+		message := fmt.Sprintf("Failed to initialise vault client for signing: %s", err)
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorVaultInit", message)
+		return err
+	}
+
+	duration, err := pki.DurationFromCertificateSigningRequest(csr)
+	if err != nil {
+		message := fmt.Sprintf("Failed to parse requested duration: %s", err)
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorParseDuration", message)
+		util.CertificateSigningRequestSetFailed(csr, "ErrorParseDuration", message)
+		_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	certPEM, caPEM, err := client.Sign(csr.Spec.Request, duration)
+	if err != nil {
+		message := fmt.Sprintf("Vault failed to sign: %s", err)
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorSigning", message)
+		util.CertificateSigningRequestSetFailed(csr, "ErrorSigning", message)
+		_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	log.V(logf.DebugLevel).Info("certificate issued")
+
+	// Kubernetes sub-resources, namely 'status', are separate API endpoints.
+	// We don't want to fire another re-sync of this CertificateSigningRequest
+	// before the `status.Certificate` field has been set as this will fire
+	// another sign call.
+	// Update the status.certificate first so that the sync from updating will
+	// not cause another issuance before setting the CA.
+	csr.Status.Certificate = certPEM
+	csr, err = v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+	if err != nil {
+		message := "Error updating certificate"
+		v.recorder.Eventf(csr, corev1.EventTypeWarning, "ErrorUpdate", "%s: %s", message, err)
+		return err
+	}
+
+	if csr.Annotations == nil {
+		csr.Annotations = make(map[string]string)
+	}
+	csr.Annotations[experimentalapi.CertificateSigningRequestCAAnnotationKey] = base64.StdEncoding.EncodeToString(caPEM)
+	_, err = v.certClient.Update(ctx, csr, metav1.UpdateOptions{})
+	if err != nil {
+		message := fmt.Sprintf("Error setting %q", experimentalapi.CertificateSigningRequestCAAnnotationKey)
+		v.recorder.Eventf(csr, corev1.EventTypeWarning, "ErrorCAUpdate", "%s: %s", message, err)
```

Comment:
question: why do we send an event here? We know that this is a transient error, we don't need to warn the user since we know this error will be fixed within a second.

Reply:
See above.